File: bench/cpu/sorted_array.cpp
#include <benchmark/benchmark.h>

#include <algorithm>

std::vector<int> MakeArr() {
  srand(time(NULL));

  std::vector<int> arr(1'000'000);
  for (auto& x : arr) {
    x = rand() % 256;
  }

  return arr;
}

// Force a real branch by making the body opaque to the optimizer
static void BM_UnsortedArray(benchmark::State& state) {
  auto arr = MakeArr();

  for (auto _ : state) {
    int sum = 0;
    for (auto i = 0u; i < arr.size(); ++i) {
      if (arr[i] > 128) {
        asm volatile("" ::: "memory");
        sum += arr[i];
      }
    }
    benchmark::DoNotOptimize(sum);
  }
}

static void BM_UnsortedArrayNoBranch(benchmark::State& state) {
  auto arr = MakeArr();

  for (auto _ : state) {
    int sum = 0;
    for (auto i = 0u; i < arr.size(); ++i) {
      sum += arr[i] * (arr[i] > 128);
    }
    benchmark::DoNotOptimize(sum);
  }
}

static void BM_UnsortedArrayNoBranch2(benchmark::State& state) {
  auto arr = MakeArr();

  for (auto _ : state) {
    int sum[2] = {0, 0};
    for (auto i = 0u; i < arr.size(); ++i) {
      sum[arr[i] > 128] += arr[i];
    }
    benchmark::DoNotOptimize(sum[true]);
  }
}

static void BM_SortedArray(benchmark::State& state) {
  auto arr = MakeArr();
  std::sort(arr.begin(), arr.end());

  for (auto _ : state) {
    int sum = 0;
    for (auto i = 0u; i < arr.size(); ++i) {
      if (arr[i] > 128) {
        asm volatile("" ::: "memory");
        sum += arr[i];
      }
    }
    benchmark::DoNotOptimize(sum);
  }
}

BENCHMARK(BM_UnsortedArray);
BENCHMARK(BM_UnsortedArrayNoBranch);
BENCHMARK(BM_UnsortedArrayNoBranch2);
BENCHMARK(BM_SortedArray);

BENCHMARK_MAIN();

File: bench/cpu/sorted_array.zig
const std = @import("std");

const N = 1_000_000;

fn makeArr() ![]i32 {
    const alloc = std.heap.page_allocator;
    const arr = try alloc.alloc(i32, N);
    var prng = std.Random.DefaultPrng.init(42);
    const rand = prng.random();
    for (arr) |*x| x.* = @intCast(rand.intRangeAtMost(u8, 0, 255));
    return arr;
}

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

pub fn main() !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    const unsorted = try makeArr();
    defer std.heap.page_allocator.free(unsorted);

    const sorted = try std.heap.page_allocator.alloc(i32, N);
    defer std.heap.page_allocator.free(sorted);
    @memcpy(sorted, unsorted);
    std.mem.sort(i32, sorted, {}, std.sort.asc(i32));

    // Unsorted with branch
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i32 = 0;
            for (unsorted) |x| {
                if (x > 128) sum +%= x;
            }
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_UnsortedArray", total, iters);
    }

    // Unsorted branchless
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i32 = 0;
            for (unsorted) |x| {
                sum +%= x * @as(i32, @intFromBool(x > 128));
            }
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_UnsortedArrayNoBranch", total, iters);
    }

    // Sorted with branch
    {
        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i32 = 0;
            for (sorted) |x| {
                if (x > 128) sum +%= x;
            }
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_SortedArray", total, iters);
    }
}
